opsi/manager/pipeline.py

```diff
@@ -24,12 +27,138 @@ class Connection(NamedTuple):
 Links = Dict[str, Connection]
 
 
+class SingleRunPerformance(NamedTuple):
+    nodes: Dict[str, float]
+
+
+class Performance:
+    # All floats here are time durations measured in seconds
+    __slots__ = ("nodes", "node_types", "pipeline", "sum_nodes", "current_run")
+
+    def __init__(self, nodes: Dict[str, "Node"]):
+        self.nodes: Dict[str, List[float]] = {id: [] for id in nodes.keys()}
+        self.node_types: Dict[str, str] = {
+            id: node.func_type.type for id, node in nodes.items()
+        }
+        self.pipeline: List[float] = []
+        self.sum_nodes: List[float] = []
+        self.current_run: SingleRunPerformance = None
+
+    def new_run(self):
+        self.current_run = SingleRunPerformance(nodes=dict.fromkeys(self.nodes.keys()))
+
+    def log_node_run(self, id, time):
+        self.current_run.nodes[id] = time
+
+    def finalize_run(self, pipeline_perf):
+        # filter out None and zero
+        self.sum_nodes.append(math.fsum(filter(None, self.current_run.nodes.values())))
+
+        for id, data in self.nodes.items():
+            entry = self.current_run.nodes[id]
+            if entry is not None:
+                data.append(entry)
+
+        self.pipeline.append(pipeline_perf)
+
+        self.current_run = None
+
+    def ensure_consistency(self):
```

Comment:
Maybe there should be a pipeline changed hook here to toss out old data?

Reply:
Why? Toss out what? The only Function using this hook is h264.

Reply:
Basically reset the profile if a new pipeline is loaded. This is a non-issue though because of the clear function.